python/modules/IcePy/ObjectAdapter.cpp

```diff
@@ -449,37 +447,50 @@ adapterWaitForHold(ObjectAdapterObject* self, PyObject* args)
     //
     if(PyThread_get_thread_ident() == _mainThreadId)
     {
-        IceUtil::Monitor<IceUtil::Mutex>::Lock sync(*self->holdMonitor);
+        std::lock_guard lock(*self->holdMutex);
 
         if(!self->held)
         {
-            if(self->holdThread == 0)
+            if(self->holdFuture == nullptr)
             {
-                AdapterInvokeThreadPtr t = new AdapterInvokeThread(*self->adapter,
-                                                                   &Ice::ObjectAdapter::waitForHold,
-                                                                   *self->holdMonitor, self->held);
-                self->holdThread = new AdapterInvokeThreadPtr(t);
-                t->start();
+                self->holdFuture = new std::future<void>();
+                *self->holdFuture = std::async(std::launch::async,
+                                               [&self]
+                                               {
+                                                   (*self->adapter)->waitForHold();
+                                               });
             }
 
-            bool done;
             {
                 AllowThreads allowThreads; // Release Python's global interpreter lock during blocking calls.
-                done = (*self->holdMonitor).timedWait(IceUtil::Time::milliSeconds(timeout));
+                if(self->holdFuture->wait_for(std::chrono::milliseconds(timeout)) == std::future_status::timeout)
+                {
+                    PyRETURN_FALSE;
+                }
             }
 
-            if(!done)
+            self->held = true;
+            try
             {
-                PyRETURN_FALSE;
+                self->holdFuture->get();
+            }
+            catch(const Ice::Exception&)
+            {
+                *self->holdException = current_exception();
             }
         }
 
         assert(self->held);
-
-        Ice::Exception* ex = (*self->holdThread)->getException();
-        if(ex)
+        if(*self->holdException)
         {
-            setPythonException(*ex);
+            try
+            {
+                std::rethrow_exception(*self->holdException);
+            }
+            catch(const Ice::Exception& ex)
```

Comment:
Should we catch std::exception is all this code?

Reply:
I don't think we need to. I'd also prefer to keep the main logic for this PR.